Build Python 3.4 wheels

mgedmin opened the pull request:
Fixes https://github.com/jonparrott/cmarkgfm/issues/2.

mgedmin committed:
Build Python 3.4 wheels

Fixes https://github.com/jonparrott/cmarkgfm/issues/2.

mgedmin committed:
Skip uploading if PyPI credentials not available

When Appveyor tests pull requests it quite sensibly doesn't give them access to secure variables.  This makes twine upload stop and ask the user for input, which makes the whole Appveyor job hang until it times out 60 minutes later.

mgedmin committed:
Try the mingwpy toolchain on Python 3.4

mgedmin committed:
Try to use mingwpy, harder

mgedmin committed:
I kan't spell

anthrotype commented on appveyor.yml:
```diff
@@ -48,14 +52,14 @@ install:
     - python -m pip install %BUILD_DEP%
     # Build wheel
-    - python -m pip wheel --wheel-dir=wheelhouse --no-binary=%PACKAGE_NAME% %PROJECT_SPEC%
+    - python -m pip wheel --wheel-dir=wheelhouse --no-binary=%PACKAGE_NAME% %PROJECT_SPEC% %EXTRA_BUILD_OPTIONS%
```
I don't think `pip wheel` can take a `--compiler` option. `bdist_wheel` does for sure.
I think pip wheel has a `--build-option <options>` that takes a string of bdist_wheel options.
You may well just call bdist_wheel directly. I don't see much gain from calling pip wheel here

mgedmin committed:
Wrap the build option in --build-option

mgedmin committed:
Try another hack

It's a shame python setup.py bdist_wheel doesn't accept --compiler=....

Let's try the hack suggested by https://pip.pypa.io/en/stable/reference/pip_wheel/#customising-the-build.  If this also fails, we might have to create a pydistutils.cfg.

anthrotype commented:
sorry I misled you. I thought bdist_wheel would take a --compiler option, but you're right it doesn't.
the way I'm using mingwpy in this other project of mine is by passing the `DISTUTILS_COMPILER=mingw32` environment variable.
https://github.com/googlei18n/compreffor/blob/142ddbee7a37bc0015d3a39137f092a073c46374/tox.ini#L13
it's been long time so I had forgotten

anthrotype commented:
actually no, forget what I said.
That `DISTUTILS_COMPILER` variable is another hack of mine, requires you to change the setup.py itself
https://github.com/googlei18n/compreffor/blob/142ddbee7a37bc0015d3a39137f092a073c46374/setup.py#L42-L44

mgedmin committed:
Seriously, what is wrong with me today

anthrotype commented:
yeah, I think the only solution is the --global-option thing you're doing (besides the distutils config file)

mgedmin committed:
Literally crying right now

anthrotype commented:
I think `pip wheel` command will attempt to build everything from source (ignoring the --no-binary option) when you pass --global-option, that's why it's trying (and failing?) to compile cffi wheel from source.
Try to pass --no-deps to pip wheel to only build the cmarkgfm

anthrotype commented:
see https://ci.appveyor.com/project/jonparrott/cmarkgfm-wheels/build/1.0.26/job/0vlclvw48d22pmvw#L841

```
no binaries permitted for cffi
```

even though cffi does provide wheels for windows py34, both 32 and 64 bit.

the reason is pip wheel recompiles everything from source when a global build option is passed, like in this case. You should pass --no-deps so it won't attempt to rebuild cffi from source (which seems to make assumptions about the compiler type and doesn't seem to work with mingw32).

Also, the only way I could force build_cmark.py `_compiler_type` to detect "mingw32" -- given it doesn't seem to have access to the setup.py command line options -- was to modify one of the config files (either the setup.cfg in the cmark source directory, or the user's `%HOME%/pydistutils.cfg`, or the site-specific `venv/Lib/distutils/distutils.cfg`) and add a `[build] compiler=mingw32` section to it.

Finally, we need tihs https://github.com/jonparrott/cmarkgfm/pull/4 to be merged and released first, for this wheel-building repository to work.
Otherwise, even if we convince the cffi builder that the compiler type is mingw32, that `if` statement in  the currently released build_cmark.py doesn't handle the case when compiler type is neither "unix" nor "msvc" leading to some NameError later on.

jonparrott commented:
I *think* this will work once I release a new cmarkgfm. Worst case, 3.4 wheels don't build.
